cmd/snap-bootstrap/partition/encrypt.go

```diff
@@ -0,0 +1,144 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package partition
+
+import (
+	"bytes"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+var (
+	tempFile = ioutil.TempFile
+)
+
+// The LUKS master key is 64 bytes long
+const (
+	masterKeySize = 64
+)
+
+// EncryptionKey holds the LUKS master key.
+type EncryptionKey [masterKeySize]byte
+
+func NewEncryptionKey() (EncryptionKey, error) {
+	var key EncryptionKey
+	_, err := rand.Read(key[:])
+	// On return, n == len(b) if and only if err == nil
+	return key, err
+}
+
+// Store writes the LUKS master key in the location specified by filename.
+func (key EncryptionKey) Store(filename string) error {
+	// XXX: must provision the TPM, generate and store the lockout authorization,
+	// and seal the key. Currently we're just storing the unprocessed data.
+	if err := ioutil.WriteFile(filename, key[:], 0600); err != nil {
+		return fmt.Errorf("cannot store key file: %v", err)
+	}
+
+	return nil
+}
+
+// EncryptedDevice represents a LUKS-backed encrypted block device.
+type EncryptedDevice struct {
+	Node          string
+	encryptedNode string
+	name          string
+}
+
+func NewEncryptedDevice(part *DeviceStructure, name string) *EncryptedDevice {
+	dev := &EncryptedDevice{
+		// A new block device is used to access the encrypted data
+		Node:          fmt.Sprintf("/dev/mapper/%s", name),
+		encryptedNode: part.Node,
+		name:          name,
+	}
+
+	return dev
+}
+
+// Encrypt creates an encrypted device in the existing partition using the
+// specified key.
+func (dev *EncryptedDevice) Encrypt(key EncryptionKey) error {
+	tempKeyFile, err := tempFile("", "enc")
+	if err != nil {
+		return err
+	}
+	defer wipe(tempKeyFile.Name())
+
+	// XXX: Ideally we shouldn't write this key, but cryptsetup
+	// only reads the master key from a file.
+	if _, err := tempKeyFile.Write(key[:]); err != nil {
+		return fmt.Errorf("cannot create key file: %s", err)
+	}
+
+	if err := cryptsetupFormat(tempKeyFile.Name(), dev.encryptedNode); err != nil {
+		return fmt.Errorf("cannot format encrypted device: %v", err)
+	}
+	if err := cryptsetupOpen(tempKeyFile.Name(), dev.encryptedNode, dev.name); err != nil {
```

Comment:
Only past that point dev.Node is available. Maybe w should have a something like this:
```
func Encrypt(dev DeviceStructure) error { 
    /* calls cryptsetup luksFormat */ 
}

type EncryptedDevice struct {
    parent *DeviceStructure
    Node string
}

func OpenEncryptedDevice(dev DeviceStructure) (EncryptedDevice, error) { 
     /* calls cryptsetup open */ 
}

func (e *EncryptedDevice) Close() error {
    /* calls cryptsetup close */
}

func (e *EncryptedDevice) AsDevice() DeviceStructure {
    return DeviceStructure{
        VolumeStructure: e.parent.VolumeStructure,
        Node: e.Node
    }
}
```

Reply:
The proposed abstraction is closer to what actually happens on partition level, requires less state data and should be clearer to follow. I'll update the code to use it.

Reply:
After further consideration, it became clear that if we consider an operation safe to be done once, then we can do it separately for format and open. Conversely, if we think that doing it multiple times isn't safe, we shouldn't use it in the first time, so I'll implement it in the cleaner way as it was originally suggested.